### PR TITLE
Sync Unit Tests: Use mocked themes instead of assumed installed themes

### PR DESCRIPTION
**Pull request by kraftbj**
WordPress doesn't ship with all of the default themes. Our unit test expects the SVN checkout of WordPress, which does. Our Docker environment, though, installs the stock version of WordPress with a limited number of themes.

Currently, if setting up a fresh Docker env, our unit tests will fail because we try to mock update non-existent themes.

#### Changes proposed in this Pull Request:
* Update the theme update tests to use mocked themes.
* Include mocked themes.

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* Completely clean out Docker and run the phpunit tests.
* `yarn docker:clean && yarn docker:up` and `yarn docker:phpunit --filter=test_update_theme`

#### Proposed changelog entry for your changes:
* n/a

### CODE REVIEW

**Comment by mdbitz** [comment]
Skip the test if the themes are not present.

I'm not sure about this last bit. Should we expect the themes to be installed and fail? What happens in the future when neither 2019 and 2020 are pre-installed? Will we fix this test?

--

I don't think skipping it is the best approach. There are 2 ways forward that I can see.

1) Instead of skip trigger a failure that the test theme is no longer available. Please update the unit test
 -- counter is that we'll still have to update the tests when they start to fail
2) We update the test to be dynamic based on wp_get_themes returned values. 
 -- issue with this is that we are reliant on non-static data so the test could pass when an issue exists.

**Comment by kraftbj** [comment]
@mdbitz Hmm, we have some tests that install a fake plugin to ensure we have reliable things to interact with. What about doing that here? A couple empty themes with a style.css?

**Comment by mdbitz** [comment]
Definitely, I'll update the test today to go this approach and make sure it also cleans up after itself :)

**Comment by mdbitz** [comment]
@kraftbj I took some test logic from the core tests. So we know have 2 dummy themes in our files folder. Pre running tests we copy them to the themes folder than after tests remove them.

We still use old themes as you can see in setUp. However we need to tackle them separately, my initial attempts cause failures as we need to have theme support usage and other elements.

Let me know what you think.

**Comment by mdbitz** [comment]
@kraftbj or @bisko let me know your thoughts. All tests pass now

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 1]
```diff
@@ -34,6 +34,65 @@ function theme_info() {
...
+	 */
+	public static function tearDownAfterClass() {
+		parent::tearDownAfterClass();
+
+		$themes = array(
+			'theme-file-sync-parent',
+			'theme-file-sync-child',
+		);
+
+		// Copy themes from tests/php/files/ to wp-content/themes.
```
Comment (anomiex) [inline tests/php/sync/test_class.jetpack-sync-themes.php:83]:
```suggestion
		// Remove themes previously copied from tests/php/files/ to wp-content/themes.
```

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 2]
```diff
@@ -34,6 +34,65 @@ function theme_info() {
...
 
+	/**
+	 * Dummy Themes.
+	 *
+	 * @var string[]
+	 */
+	protected $themes = array(
+		'theme-file-sync-parent',
+		'theme-file-sync-child',
+	);
```
Comment (anomiex) [inline tests/php/sync/test_class.jetpack-sync-themes.php:45]: You don't seem to ever use this. Why not make it static, and then use it in `setUpBeforeClass()` and `tearDownAfterClass()` below?

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 3]
```diff
@@ -262,22 +320,24 @@ public function test_update_themes_sync() {
...
-		//Not testing versions since they are subject to change
-		$this->assertEquals( 'Twenty Seventeen', $themes['twentyseventeen']['name'] );
-		$this->assertEquals( 'https://wordpress.org/themes/twentyseventeen/', $themes['twentyseventeen']['uri'] );
-		$this->assertEquals( 'twentyseventeen', $themes['twentyseventeen']['stylesheet'] );
-		$this->assertEquals( 'Twenty Sixteen', $themes['twentysixteen']['name'] );
-		$this->assertEquals( 'https://wordpress.org/themes/twentysixteen/', $themes['twentysixteen']['uri'] );
-		$this->assertEquals( 'twentysixteen', $themes['twentysixteen']['stylesheet'] );
+		$themes     = $event_data->args[0];
+
+		// Not testing versions since they are subject to change.
```
Comment (anomiex) [inline tests/php/sync/test_class.jetpack-sync-themes.php:325]: Is this still true now that we're supplying our own themes?

Same on line 351 below.

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 4]
```diff
@@ -34,6 +34,65 @@ function theme_info() {
...
+			}
+
+			rmdir( $dest_dir );
+		}
+
+	}
+
 	public function setUp() {
 		parent::setUp();
 		$themes      = array( 'twentyten', 'twentyeleven', 'twentytwelve', 'twentythirteen', 'twentyfourteen' );
```
Comment (anomiex) [inline tests/php/sync/test_class.jetpack-sync-themes.php:98]: What's going on here? Can we remove the random switching to a theme that doesn't exist in the installation anymore?

Locally if I just set `$this->theme = 'doesnotexist';` the tests still seem to pass.

**Comment by anomiex** [comment]
Seems like the PR title and summary may need updating for the changed direction.

**Comment by kraftbj** [comment]
> Seems like the PR title and summary may need updating for the changed direction.

Done!

**Comment by anomiex** [comment]
Still some comments from yesterday that could use a reply.

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 3 above)
Comment (mdbitz) [reply tests/php/sync/test_class.jetpack-sync-themes.php:325]: more appropriate action would be a to-do or add check on version but that is not in scope.

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 2 above)
Comment (mdbitz) [reply tests/php/sync/test_class.jetpack-sync-themes.php:45]: :thumbsup: had intentions to use it but didn't get around to it.

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 3 above)
Comment (anomiex) [reply tests/php/sync/test_class.jetpack-sync-themes.php:325]: I'd be fine with changing the comment to a to-do.

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 2 above)
Comment (mdbitz) [reply tests/php/sync/test_class.jetpack-sync-themes.php:45]: Updated.

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 4 above)
Comment (mdbitz) [reply tests/php/sync/test_class.jetpack-sync-themes.php]: removed in latest update.

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 3 above)
Comment (mdbitz) [reply tests/php/sync/test_class.jetpack-sync-themes.php]: updated both instances of this comment to a TODO

**Comment by mdbitz** [comment]
Updates are showing successful tests locally. If any issue with tests in Travis, lets then punt to 9.3 as this isn't tied to the release in any way.

**Comment by mdbitz** [comment]
Updating to 9.3 as multisite run though is being flakey.

**Review by anomiex — DISMISSED** [review]

**Comment by kraftbj** [comment]
There's a PHP 7.4 test failure:

```
There was 1 failure:

1) WP_Test_Jetpack_Sync_Themes::test_network_enable_disable_theme_sync
Failed asserting that false is true.

/tmp/wordpress-master/src/wp-content/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-themes.php:240
/tmp/wordpress-master/src/wp-content/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-themes.php:192
```

**Comment by kraftbj** [comment]
All tests pass now. @mdbitz Was there anything else outstanding on this? You marked it In Progress, so I don't want to move it into the review and merge pipeline without your nod.

**Review by mdbitz — APPROVED** [review]
No concerns from me, I had marked this In Progress to come back to the test issue. Lets merge and get his resolved.

**Pull request merged**